Moving archiver to dfs-datastore and added PailRecord type dependency…

gsriram7 opened the pull request:
… to PailMigrate

@sriram-r @ashwanthkumar

sriram-r commented:
We shouldn't call this archiver anymore. It only helps do pail spec migration.

ashwanthkumar commented on dfs-datastores/src/main/scala/com/indix/pail/PailMigrate.scala:
```diff
@@ -47,7 +47,7 @@ class PailMigrate extends Configured with Tool {
-    * The output location will be (vertically) partitioned based on the `specClass` parameter
+    * The output location will be (vertically) partitioned based on the `targetSpecClass` parameter
     * */
     val outputDir = args(1)
```
@gsriram7 I think we discussed this earlier, can we please use scalding-args?

gsriram7 replied:
We did talk about using scalding but I thought you are convinced on the reasons why we didn't use it. Read https://github.com/ind9/dfs-datastores/pull/13

ashwanthkumar replied:
That was for writing the job using Scalding right? Here I'm suggesting to use just scalding-args. That's a module with just 2 classes that helps you in Arg parsing. It is a small and neat library.

gsriram7 replied:
Oh scalding-args, yes it will be appropriate to use it.

ashwanthkumar commented:
With lot of active changes on this repo, @sriram-r @ganeshpraburavi @gsriram7 Do you guys think is this the right time to create a separate module for these jobs and convert 'em to an utility? That way I could push these to the environment (hadoop clusters) as part of Chef and we don't need each repo have it's own version of PailConsolidate or PailArchiver. 

As a dev / ops / go pipeline it would be great just ssh into the machine and do
```
$ pail consolidate s3n://indix-masterdata/prices/new/20151105/12/am
or
$ pail move s3n://indix-masterdata/prices/new/20151105/12/am s3n://indix-masterdata/prices/master
or
$ pail copy s3n://indix-masterdata/prices/new/20151105/12/am s3n://indix-masterdata/prices/master
```
Thoughts?

ashwanthkumar commented on dfs-datastores/src/main/scala/com/indix/pail/PailMigrate.scala:
```diff
@@ -130,8 +133,8 @@ object PailMigrate  {
     override def configure(jobConf: JobConf): Unit = {
-      recordClass = Class.forName(jobConf.get("pail.migrate.record"))
       outputPailStructure = Utils.getObject(jobConf, OUTPUT_STRUCTURE).asInstanceOf[PailStructure[Any]]
+      recordClass = Class.forName(jobConf.get("pail.migrate.record"))
```
@gsriram7 Why do we need this `recordClass`?

gsriram7 replied:
Sorry we had that for some other reason which isn't relevant anymore. Will remove it..

sriram-r commented:
Agree. Ideally, the pail-utils should come out as a module. But, lets not
hold the merge waiting for this. We should be playing that separately.

On Thu, Nov 5, 2015 at 9:52 PM, Ashwanth Kumar <notifications@github.com>
wrote:

> With lot of active changes on this repo, @sriram-r
> <https://github.com/sriram-r> @ganeshpraburavi
> <https://github.com/ganeshpraburavi> @gsriram7
> <https://github.com/gsriram7> Do you guys think is this the right time to
> create a separate module for these jobs and convert 'em to an utility? That
> way I could push these to the environment (hadoop clusters) as part of Chef
> and we don't need each repo have it's own version of PailConsolidate or
> PailArchiver.
>
> As a dev / ops / go pipeline it would be great just ssh into the machine
> and do
>
> $ pail consolidate s3n://indix-masterdata/prices/new/20151105/12/am
> or
> $ pail move s3n://indix-masterdata/prices/new/20151105/12/am s3n://indix-masterdata/prices/master
> or
> $ pail copy s3n://indix-masterdata/prices/new/20151105/12/am s3n://indix-masterdata/prices/master
>
> Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ind9/dfs-datastores/pull/16#issuecomment-154110853>.
>

ashwanthkumar commented:
@sriram-r If you think we'll play it outside of a context then it is fine.

gsriram7 commented:
@ashwanthkumar @sriram-r 
Sorry for the delay, i incorporated the changed which made sense to me. 
Introduced scalding-args in PailMigrate. 
Manually tested both PailMigrate and IxPailConsolidator.
Please share your thoughts on this.

sriram-r commented on dfs-datastores/src/main/scala/com/indix/pail/PailMigrate.scala:
```diff
@@ -29,40 +30,31 @@ class PailMigrate extends Configured with Tool {
-      println("record_type - Fully qualified class name of the record type. Eg: com.indix.models.ProductRecordPail")
-      println("keep_source_files - set this to true to keep the source files. This is optional, Defaults to false, which means, source files will be removed if the job succeeds")
+  override def run(arguments: Array[String]): Int = {
+    if (arguments.length < 4) {
```
Selva - The idea behind using scalding-args is to allow it to do the heavy lifting of mandatory vs optional and help messages. So, this whole block should go away.

gsriram7 replied:
Yes I understood its usage as Ashwanth mentioned it last time hence i removed few comments to describe the format of each argument. I thought there should be a helper message to assist users as we want the users to explicitly mention fqcn

sriram-r commented on dfs-datastores/src/main/scala/com/indix/pail/PailMigrate.scala:
```diff
@@ -29,40 +30,31 @@ class PailMigrate extends Configured with Tool {
     val recordClass = Class.forName(recordType)
 
-    val keepSourceFiles = Option(args(4)).exists(_ equals "true")
+    val keepSourceFiles = Option(args("keep-source")).exists(_ equals "true")
```
this could be better done as `args.getOrElse` itself.

gsriram7 replied:
sure

sriram-r commented on dfs-datastores/src/main/scala/com/indix/pail/PailMigrate.scala:
```diff
@@ -73,34 +65,32 @@ class PailMigrate extends Configured with Tool {
     FileOutputFormat.setOutputPath(jobConf, new Path(outputDir))
 
-    Utils.setObject(jobConf, PailMigrate.OUTPUT_STRUCTURE, pailStructure)
+    com.backtype.support.Utils.setObject(jobConf, PailMigrate.OUTPUT_STRUCTURE, targetPailStructure)
```
Can't we push this `com.backtype.support` to import?

gsriram7 replied:
sure

ashwanthkumar commented on dfs-datastores/src/main/scala/com/indix/pail/PailMigrate.scala:
```diff
@@ -73,34 +55,32 @@ class PailMigrate extends Configured with Tool {
 
-    jobConf.set("pail.migrate.record",recordType)
-
     val job = new JobClient(jobConf).submitJob(jobConf)
```
@gsriram7 isn't submitJob a blocking call?

ashwanthkumar commented on dfs-datastores/src/main/scala/com/indix/pail/PailMigrate.scala:
```diff
@@ -25,44 +26,25 @@ class PailMigrate extends Configured with Tool {
+    val targetSpecClass = args("target-pail-spec")
 
-    val recordType = args(3)
+    val recordType = args("record-type")
```
@gsriram7 Do we need this param? I see in the mapper while we read the serialized object we're typecasting it to `PailStructure[Any]`?
